### PR TITLE
Implementation of Block-Lanczos procedure

### PR DESCRIPTION
**Pull request by mfherbst**
Less useful for standard ADC use cases, but very useful for debugging and for special applications (e.g. large number of eigenpairs).

## TODO
- [x] Shift-and-invert matrix
- [x] Docstrings

### CODE REVIEW

File: adcc/solver/LanczosIterator.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+        # To be initialised by __iter__
+        self.lanczos_subspace = []
+        self.alphas = []  # Diagonal matrix block of subspace matrix
+        self.betas = []   # Side-diagonal matrix blocks of subspace matrix
+        self.residual = guesses
+        self.n_iter = 0
+        self.n_applies = 0
+
+    def __iter__(self):
```
Comment (maxscheurer) [inline adcc/solver/LanczosIterator.py:80]: Where are these functions actually called? 😄 
I'm a bit overwhelmed by the Python magic probably...

Comment (mfherbst) [reply adcc/solver/LanczosIterator.py:80]: Well basically this is an iterator that "iterates" through the Lanczos subspace. They are implicitly called in the loop "for subspace in iterator" in the lanczos.py file and setup and advance the state.

Comment (maxscheurer) [reply adcc/solver/LanczosIterator.py:80]: That's cool 👍

**Review by maxscheurer — APPROVED** [review]
I cannot comment on this from an "algorithm"/math point of view, but the code looks very clean, well documented, etc. 😄

**Comment by mfherbst** [comment]
No worries. I hope it's doing good. I first implemented it in Julia where I really treated it super badly when it comes to numerics, then I transferred it to python and checked every step. It might still be wrong but the final residuals will always tell you, so at worst it just does not find the eigenvalues, but wrong results will not go unnoticed.

Thanks for looking at the code and documentation!

**Pull request merged**